File: backend/src/main/java/com/labex/labexagent/run/AgentRunState.java
package com.labex.labexagent.run;

import java.util.Locale;

public enum AgentRunState {
    QUEUED,
    PREPARING,
    RUNNING,
    WAITING_APPROVAL,
    WAITING_USER,
    WAITING_WORKSPACE,
    WAITING_ENVIRONMENT,
    RECOVERING,
    RETRYING,
    CANCELLING,
    CANCELLED,
    FAILED,
    COMPLETED;

    public String persistedStatus() {
        return name().toLowerCase(Locale.ROOT);
    }

    public static AgentRunState fromPersistedStatus(String value) {
        String normalized = value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
        if ("pending".equals(normalized)) {
            return QUEUED;
        }
        for (AgentRunState state : values()) {
            if (state.persistedStatus().equals(normalized)) {
                return state;
            }
        }
        throw new IllegalArgumentException("Unknown agent run state: " + value);
    }
}

File: frontend/src/composables/agentRunState.js
const TERMINAL_RUN_STATES = new Set(['completed', 'failed', 'cancelled'])
const RECOVERABLE_RUN_STATES = new Set([
  'waiting_approval',
  'waiting_user',
  'waiting_workspace',
  'waiting_environment',
  'retrying',
  'retry_backoff'
])

export function normalizeAgentRunState(value) {
  return String(value || '').trim().toLowerCase()
}

export function isTerminalAgentRunState(value) {
  return TERMINAL_RUN_STATES.has(normalizeAgentRunState(value))
}

export function isRecoverableAgentRunState(value) {
  return RECOVERABLE_RUN_STATES.has(normalizeAgentRunState(value))
}
